File: scripts/release/prepare_release_win.py
import os, pathlib, re, sys

if len(sys.argv) < 2:
    print("usage: python prepare_release_win.py <new-version>")
    sys.exit(1)

new = sys.argv[1]
if not re.match(r"^\d+\.\d+\.\d+$", new):
    print(f"error: '{new}' is not a plain X.Y.Z version")
    sys.exit(1)

# read old version from Cargo.toml
p = pathlib.Path("Cargo.toml")
text = p.read_text(encoding='utf-8')
m = re.search(r'^version = "([^"]+)"', text, re.MULTILINE)
if not m:
    print("error: could not find version in Cargo.toml")
    sys.exit(1)
old = m.group(1)

if old == new:
    print(f"workspace is already at {new}; nothing to bump")
    sys.exit(0)

print(f"Bumping {old} -> {new}")
old_re = re.escape(old)

def bump(path, pattern, repl, minimum):
    p = pathlib.Path(path)
    text = p.read_text(encoding='utf-8')
    out, n = re.subn(pattern, repl, text, flags=re.MULTILINE)
    if n < minimum:
        sys.exit(f"error: expected >= {minimum} replacement(s) in {path}, made {n}")
    p.write_text(out, encoding='utf-8')
    print(f"  {path}: {n} replacement(s)")

# 1) Workspace version.
bump("Cargo.toml", rf'^version = "{old_re}"$', f'version = "{new}"', 1)

# 2) Internal helpofai-* dependency pins in every crate manifest.
total = 0
for manifest in sorted(pathlib.Path("crates").glob("*/Cargo.toml")):
    text = manifest.read_text(encoding='utf-8')
    out, n = re.subn(
        rf'(helpofai-[a-z0-9-]+\s*=\s*\{{[^}}]*version = ")[0-9.]+(")',
        rf"\g<1>{new}\g<2>",
        text,
    )
    if n:
        manifest.write_text(out, encoding='utf-8')
        print(f"  {manifest}: {n} pin(s)")
        total += n
if total == 0:
    sys.exit("error: no internal dependency pins were bumped — wrong old version?")

# 3) npm wrappers.
bump(
    "npm/helpofai/package.json",
    rf'("(?:version|helpofaiBinaryVersion)": ")[0-9.]+(")',
    rf"\g<1>{new}\g<2>",
    2,
)
bump(
    "npm/runtime-sdk/package.json",
    rf'("(?:version)": ")[0-9.]+(")',
    rf"\g<1>{new}\g<2>",
    1,
)

# 4) README install-tag examples (all translations).
# Using a generic regex so we match even if the readme version was out of sync (e.g. 0.8.70).
for readme in ["README.md", "README.zh-CN.md", "README.ja-JP.md", "README.vi.md"]:
    if pathlib.Path(readme).exists():
        bump(readme, rf"--tag v[0-9.]+\b", f"--tag v{new}", 1)

print("Bumping complete.")

File: scripts/release/prepare_release_win.js
const fs = require('fs');
const path = require('path');

const args = process.argv.slice(2);
if (args.length < 1) {
    console.error('usage: node prepare_release_win.js <new-version>');
    process.exit(1);
}

const newVersion = args[0];
if (!/^\d+\.\d+\.\d+$/.test(newVersion)) {
    console.error(`error: '${newVersion}' is not a plain X.Y.Z version`);
    process.exit(1);
}

// 1. Cargo.toml workspace
let cargo = fs.readFileSync('Cargo.toml', 'utf8');
const oldVersionMatch = cargo.match(/^version = "([^"]+)"/m);
if (!oldVersionMatch) {
    console.error('error: could not find version in Cargo.toml');
    process.exit(1);
}
const oldVersion = oldVersionMatch[1];
if (oldVersion === newVersion) {
    console.log(`workspace is already at ${newVersion}; nothing to bump`);
    process.exit(0);
}

console.log(`Bumping ${oldVersion} -> ${newVersion}`);

cargo = cargo.replace(/^version = "[^"]+"/m, `version = "${newVersion}"`);
fs.writeFileSync('Cargo.toml', cargo, 'utf8');
console.log('  Cargo.toml: version bumped');

// 2. crates/*/Cargo.toml
const cratesDir = 'crates';
fs.readdirSync(cratesDir).forEach(dir => {
    const manifestPath = path.join(cratesDir, dir, 'Cargo.toml');
    if (fs.existsSync(manifestPath)) {
        let manifest = fs.readFileSync(manifestPath, 'utf8');
        const original = manifest;
        // Bump both workspace dependency pins and version if defined
        manifest = manifest.replace(/(helpofai-[a-z0-9-]+\s*=\s*\{[^}]*version = ")[0-9.]+(",?)/g, `$1${newVersion}$2`);
        manifest = manifest.replace(/(helpflow\s*=\s*\{[^}]*version = ")[0-9.]+(",?)/g, `$1${newVersion}$2`);
        if (manifest.includes('version = "') && !manifest.includes('version = "workspace"')) {
            manifest = manifest.replace(/^version = "[^"]+"/m, `version = "${newVersion}"`);
        }
        if (manifest !== original) {
            fs.writeFileSync(manifestPath, manifest, 'utf8');
            console.log(`  ${manifestPath}: pins bumped`);
        }
    }
});

// 3. npm/helpofai/package.json & npm/runtime-sdk/package.json
const npmPaths = [
    'npm/helpofai/package.json',
    'npm/runtime-sdk/package.json'
];
npmPaths.forEach(p => {
    if (fs.existsSync(p)) {
        let pkg = fs.readFileSync(p, 'utf8');
        pkg = pkg.replace(/"(version|helpofaiBinaryVersion)": "[0-9.]+"/g, `"$1": "${newVersion}"`);
        fs.writeFileSync(p, pkg, 'utf8');
        console.log(`  ${p}: version bumped`);
    }
});

// 4. README files
const readmes = ['README.md', 'README.zh-CN.md', 'README.ja-JP.md', 'README.vi.md'];
readmes.forEach(r => {
    if (fs.existsSync(r)) {
        let text = fs.readFileSync(r, 'utf8');
        text = text.replace(/--tag v[0-9.]+\b/g, `--tag v${newVersion}`);
        fs.writeFileSync(r, text, 'utf8');
        console.log(`  ${r}: readme tags bumped`);
    }
});

console.log('Bumping complete.');
